### ISSUE TITLE
(Single) GPU support for custom Python layers?

### ISSUE THREAD

**Issue by phugen** [BVLC/caffe#5286]
This question was asked on caffe-users several times already in the past months, but nobody there posted an answer, so please bear with me.

Is it possible to use even single-GPU with custom Python layers by supplying a CUDA implementation of the _forward_ and _backward_ methods, like in the C++ layers? I've seen issues regarding "multi-GPU support in Python", so I guess it must be possible. If yes, is there any sort of guideline or example on how to "link" the Python layer with the CUDA implementations?

**Comment by williford** [comment]
Have you looked at this:
https://groups.google.com/forum/#!topic/caffe-users/ufpGi3VviVA
and the referenced link:
https://github.com/BVLC/caffe/pull/2102

Since this indicates that the answer is "not yet", I'm closing the issue. I'm posting both the question and my answer on http://stackoverflow.com [here](http://stackoverflow.com/questions/42297980/does-caffes-pythonlayer-support-gpu/42297981).

I think it is nice to link to emails referenced on caffe-users, so one of the unanswered questions is here:
https://groups.google.com/d/topic/caffe-users/uGZe4mpFWuE/discussion

**Comment by PENGUINLIONG** [comment]
I'm not sure if you are asking for availability of PythonLayer-on-GPU. If so, there might be difficulties in implementation. AFAIK, GPU-accelerated procedures cannot have sys-calls.

As caffe is a CNN _framework_, not a _library_, which calls _you_ instead of being called _by you_, it will be caffe's responsibility to check sys-calls. So it seems that another Python project (which is able to check whether there are sys-calls in scripts) has to be made. Currently, caffe is using boost to communicate with python scripts and NVCC doesn't support python compilation.

**Comment by VasLem** [comment]
Sorry for opening this again, if anyone is inter
In other words, if I am going to train a network with a Python Layer, let's say a custom loss one, it will not work with NCCL? What is the expected behavior? I tried to get the Euclidean Loss Layer, implemented [here](https://github.com/BVLC/caffe/blob/master/examples/pycaffe/layers/pyloss.py), but it seems like, after setting weight_loss:1 in the prototxt, the loss layer does not work. I tried with a different layer, a Cosine Distance Loss one, and again there is not the wanted result. The network I tested it on works with other CPP loss layers. Is there no framework for GPU-CPU interchange if one of them is not able to process the layer? Do I have to completely remove the GPU participation from the network, so that I make it work?
